Instruction count benchmark (1): Types and interfaces

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51123 Instruction count benchmark (7): Use  when collecting callgrind.
* #51122 Intermezzo: Add Timer support for callgrind repeats
* #51121 Intermezzo: quality of life tweaks to Timer. (Retain callgrind.out, change collect_baseline default, add __mul__ for FunctionCounts)
* #50610 Instruction count benchmark (6): Add frontend to drive microbenchmarks, and start README
* #50609 Instruction count benchmark (5): Add benchmark runner.
* #50608 Instruction count benchmark (4): Extend the benchmarks in both coverage and complexity.
* #50607 Instruction count benchmark (3): Enable TorchScript benchmarks
* #50606 Instruction count benchmark (2): Initial benchmark definitions.
* **#50605 Instruction count benchmark (1): Types and interfaces**
* #50604 add global_setup for C++ includes
* #50603 add utility to back test Timer

This PR sets the core interfaces the the suite uses. The entire suite leans heavily on dataclasses and enums for type safety, and it is all MyPy strict compliant. (Though I still need to set up unit tests, since it lives in `benchmarks/` rather than `torch/`.) The other main source of documentation is the README in https://github.com/pytorch/pytorch/pull/50610.

Commit:
Instruction count benchmark (1): Types and interfaces

[ghstack-poisoned]

Commit:
Update on "Instruction count benchmark (1): Types and interfaces"




[ghstack-poisoned]

Commit:
Update on "Instruction count benchmark (1): Types and interfaces"




[ghstack-poisoned]

Commit:
Update on "Instruction count benchmark (1): Types and interfaces"




[ghstack-poisoned]

swolchok commented on benchmarks/instruction_counts/core/api.py:
```diff
@@ -0,0 +1,171 @@
+    statement which is not TorchScript-able.
+
+    There are also two programming idioms in PyTorch. One is to write free form
+    code (so called "NumPy with gradients"), and the other is to organize code
```
nit: hyphenate "so-called" because it's a compound adjective

swolchok commented:
FWIW, I personally find it hard to review this diff separately from the implementations.

robieta commented:
> FWIW, I personally find it hard to review this diff separately from the implementations.

https://github.com/pytorch/pytorch/compare/gh/taylorrobie/timer_ci_prep is the overall diff if that makes it easier.

Commit:
Update on "Instruction count benchmark (1): Types and interfaces"


This PR sets the core interfaces the the suite uses. The entire suite leans heavily on dataclasses and enums for type safety, and it is all MyPy strict compliant. (Though I still need to set up unit tests, since it lives in `benchmarks/` rather than `torch/`.) The other main source of documentation is the README in https://github.com/pytorch/pytorch/pull/50610.

[ghstack-poisoned]

Commit:
Update on "Instruction count benchmark (1): Types and interfaces"


This PR sets the core interfaces the the suite uses. The entire suite leans heavily on dataclasses and enums for type safety, and it is all MyPy strict compliant. (Though I still need to set up unit tests, since it lives in `benchmarks/` rather than `torch/`.) The other main source of documentation is the README in https://github.com/pytorch/pytorch/pull/50610.

[ghstack-poisoned]
